File: impl/ocaml/src/gradients.ml
(* Gradient utilities (SPEC §10.8, "Gradients"). *)

open Ast
open Candidate
open Utilities
open Utils

(* The stops value for from-* and to-*. *)
let gradient_stops =
  "var(--tw-gradient-via-stops, var(--tw-gradient-position), var(--tw-gradient-from) var(--tw-gradient-from-position), var(--tw-gradient-to) var(--tw-gradient-to-position))"

(* The stops value set by via-*. *)
let gradient_via_stops =
  "var(--tw-gradient-position), var(--tw-gradient-from) var(--tw-gradient-from-position), var(--tw-gradient-via) var(--tw-gradient-via-position), var(--tw-gradient-to) var(--tw-gradient-to-position)"

let interpolation_methods =
  [ "srgb"; "srgb-linear"; "display-p3"; "a98-rgb"; "prophoto-rgb"; "rec2020"; "lab"; "oklab"; "xyz"; "xyz-d50"; "xyz-d65";
    "hsl"; "hwb"; "lch"; "oklch" ]

let hue_methods = [ "longer"; "shorter"; "increasing"; "decreasing" ]

let gradient_sides =
  [ ("t", "top"); ("tr", "top right"); ("r", "right"); ("br", "bottom right"); ("b", "bottom"); ("bl", "bottom left");
    ("l", "left"); ("tl", "top left") ]

(* The registrations emitted by every stop utility. *)
let gradient_registrations () =
  let open Builtin_variants in
  let transparent = Some "#0000" in
  [ property "--tw-gradient-position" None;
    property_registration "--tw-gradient-from" transparent "<color>" false;
    property_registration "--tw-gradient-via" transparent "<color>" false;
    property_registration "--tw-gradient-to" transparent "<color>" false;
    property "--tw-gradient-stops" None; property "--tw-gradient-via-stops" None;
    property_registration "--tw-gradient-from-position" (Some "0%") "<length-percentage>" false;
    property_registration "--tw-gradient-via-position" (Some "50%") "<length-percentage>" false;
    property_registration "--tw-gradient-to-position" (Some "100%") "<length-percentage>" false ]

(* Resolves an interpolation modifier. *)
let interpolation (modifier : modifier option) =
  match modifier with
  | None -> Some "in oklab"
  | Some { mkind = Arbitrary; mvalue } -> Some mvalue
  | Some { mvalue; _ } ->
      if List.mem mvalue interpolation_methods then Some ("in " ^ mvalue)
      else if List.mem mvalue hue_methods then Some ("in oklch " ^ mvalue ^ " hue")
      else None

(* Registers the gradient utilities. *)
let register u theme =
  let image shape = decl "background-image" (shape ^ "-gradient(var(--tw-gradient-stops))") in
  let positioned position modifier shape =
    match interpolation modifier with
    | None -> not_handled
    | Some method_ ->
        let value = if position = "" then method_ else position ^ " " ^ method_ in
        handled [ decl "--tw-gradient-position" value; image shape ]
  in
  let linear (c : candidate) negative legacy =
    match c.cvalue with
    | None -> not_handled
    | Some { vkind = Arbitrary; value; _ } ->
        if legacy || negative then not_handled
        else if arbitrary_type c [ "angle" ] = "angle" then positioned value c.cmodifier "linear"
        else if c.cmodifier <> None then not_handled
        else handled [ decl "background-image" ("linear-gradient(" ^ value ^ ")") ]
    | Some { value = named; _ } ->
        if has_prefix named "to-" then
          if negative then not_handled
          else
            match List.assoc_opt (after named 3) gradient_sides with
            | Some side -> positioned ("to " ^ side) c.cmodifier "linear"
            | None -> not_handled
        else if legacy || not (is_positive_integer named) then not_handled
        else
          let angle = if negative then "calc(" ^ named ^ "deg * -1)" else named ^ "deg" in
          positioned angle c.cmodifier "linear"
  in
  functional u "bg-linear" (fun c -> linear c false false);
  functional u "-bg-linear" (fun c -> linear c true false);
  functional u "bg-gradient" (fun c -> linear c false true);
  functional u "bg-radial" (fun c ->
      match c.cvalue with
      | None -> positioned "" c.cmodifier "radial"
      | Some { vkind = Arbitrary; value; _ } when c.cmodifier = None ->
          handled [ decl "--tw-gradient-position" value; image "radial" ]
      | _ -> not_handled);
  let conic (c : candidate) negative =
    match c.cvalue with
    | None -> if negative then not_handled else positioned "" c.cmodifier "conic"
    | Some { vkind = Arbitrary; value; _ } ->
        if negative || c.cmodifier <> None then not_handled else handled [ decl "--tw-gradient-position" value; image "conic" ]
    | Some { value = named; _ } ->
        if not (is_positive_integer named) then not_handled
        else
          let angle = if negative then "calc(" ^ named ^ "deg * -1)" else named ^ "deg" in
          positioned ("from " ^ angle) c.cmodifier "conic"
  in
  functional u "bg-conic" (fun c -> conic c false);
  functional u "-bg-conic" (fun c -> conic c true);
  let stop name color_variable position_variable stops =
    let color value =
      handled (gradient_registrations () @ [ decl color_variable value ] @ List.map (fun (p, v) -> decl p v) stops)
    in
    let position value = handled (gradient_registrations () @ [ decl position_variable value ]) in
    functional u name (fun c ->
        match c.cvalue with
        | None -> not_handled
        | Some { vkind = Arbitrary; value; _ } -> (
            match arbitrary_type c [ "color"; "length"; "percentage" ] with
            | "length" | "percentage" -> if c.cmodifier <> None then not_handled else position value
            | _ -> ( match as_color value c.cmodifier theme with Some r -> color r | None -> not_handled))
        | Some { value = named; _ } -> (
            match resolve_theme_color c [ "--background-color"; "--color" ] theme with
            | Some theme_color -> color theme_color
            | None ->
                if c.cmodifier <> None then not_handled
                else if has_suffix named "%" && is_positive_integer (trim_suffix named "%") then position named
                else not_handled))
  in
  stop "from" "--tw-gradient-from" "--tw-gradient-from-position" [ ("--tw-gradient-stops", gradient_stops) ];
  stop "via" "--tw-gradient-via" "--tw-gradient-via-position"
    [ ("--tw-gradient-via-stops", gradient_via_stops); ("--tw-gradient-stops", "var(--tw-gradient-via-stops)") ];
  stop "to" "--tw-gradient-to" "--tw-gradient-to-position" [ ("--tw-gradient-stops", gradient_stops) ]

File: impl/go/gradients.go
package twill

import "strings"

// Gradient utilities (SPEC §10.8, "Gradients").

// GradientStops is the stops value for from-* and to-*.
const GradientStops = "var(--tw-gradient-via-stops, var(--tw-gradient-position), var(--tw-gradient-from) var(--tw-gradient-from-position), var(--tw-gradient-to) var(--tw-gradient-to-position))"

// GradientViaStops is the stops value set by via-*.
const GradientViaStops = "var(--tw-gradient-position), var(--tw-gradient-from) var(--tw-gradient-from-position), var(--tw-gradient-via) var(--tw-gradient-via-position), var(--tw-gradient-to) var(--tw-gradient-to-position)"

var interpolationMethods = map[string]bool{
	"srgb": true, "srgb-linear": true, "display-p3": true, "a98-rgb": true, "prophoto-rgb": true,
	"rec2020": true, "lab": true, "oklab": true, "xyz": true, "xyz-d50": true, "xyz-d65": true,
	"hsl": true, "hwb": true, "lch": true, "oklch": true,
}
var hueMethods = map[string]bool{"longer": true, "shorter": true, "increasing": true, "decreasing": true}

var gradientSides = map[string]string{
	"t": "top", "tr": "top right", "r": "right", "br": "bottom right",
	"b": "bottom", "bl": "bottom left", "l": "left", "tl": "top left",
}

// GradientRegistrations returns the registrations emitted by every stop utility.
func GradientRegistrations() []Node {
	transparent := strPtr("#0000")
	return []Node{
		Property("--tw-gradient-position", nil),
		PropertyRegistration("--tw-gradient-from", transparent, "<color>", false),
		PropertyRegistration("--tw-gradient-via", transparent, "<color>", false),
		PropertyRegistration("--tw-gradient-to", transparent, "<color>", false),
		Property("--tw-gradient-stops", nil),
		Property("--tw-gradient-via-stops", nil),
		PropertyRegistration("--tw-gradient-from-position", strPtr("0%"), "<length-percentage>", false),
		PropertyRegistration("--tw-gradient-via-position", strPtr("50%"), "<length-percentage>", false),
		PropertyRegistration("--tw-gradient-to-position", strPtr("100%"), "<length-percentage>", false),
	}
}

// Interpolation resolves an interpolation modifier.
func Interpolation(modifier *Modifier) (string, bool) {
	if modifier == nil {
		return "in oklab", true
	}
	if modifier.Kind == ValueArbitrary {
		return modifier.Value, true
	}
	if interpolationMethods[modifier.Value] {
		return "in " + modifier.Value, true
	}
	if hueMethods[modifier.Value] {
		return "in oklch " + modifier.Value + " hue", true
	}
	return "", false
}

// RegisterGradientUtilities registers the gradient utilities.
func RegisterGradientUtilities(u *Utilities, theme *Theme) {
	image := func(shape string) Node {
		return Decl("background-image", shape+"-gradient(var(--tw-gradient-stops))")
	}
	positioned := func(position string, modifier *Modifier, shape string) ([]Node, CompileStatus) {
		method, ok := Interpolation(modifier)
		if !ok {
			return nil, NotHandled
		}
		value := method
		if position != "" {
			value = position + " " + method
		}
		return []Node{Decl("--tw-gradient-position", value), image(shape)}, Handled
	}

	linear := func(c *Candidate, negative, legacy bool) ([]Node, CompileStatus) {
		if c.Value == nil {
			return nil, NotHandled
		}
		if c.Value.Kind == ValueArbitrary {
			if legacy || negative {
				return nil, NotHandled
			}
			value := c.Value.Value
			t := ""
			if c.Value.DataType != nil {
				t = *c.Value.DataType
			} else {
				t = InferDataType(value, []string{"angle"})
			}
			if t == "angle" {
				return positioned(value, c.Modifier, "linear")
			}
			if c.Modifier != nil {
				return nil, NotHandled
			}
			return []Node{Decl("background-image", "linear-gradient("+value+")")}, Handled
		}
		named := c.Value.Value
		if strings.HasPrefix(named, "to-") {
			if negative {
				return nil, NotHandled
			}
			side, ok := gradientSides[named[3:]]
			if !ok {
				return nil, NotHandled
			}
			return positioned("to "+side, c.Modifier, "linear")
		}
		if legacy || !IsPositiveInteger(named) {
			return nil, NotHandled
		}
		angle := named + "deg"
		if negative {
			angle = "calc(" + named + "deg * -1)"
		}
		return positioned(angle, c.Modifier, "linear")
	}
	u.Functional("bg-linear", func(c *Candidate) ([]Node, CompileStatus) { return linear(c, false, false) }, nil)
	u.Functional("-bg-linear", func(c *Candidate) ([]Node, CompileStatus) { return linear(c, true, false) }, nil)
	u.Functional("bg-gradient", func(c *Candidate) ([]Node, CompileStatus) { return linear(c, false, true) }, nil)

	u.Functional("bg-radial", func(c *Candidate) ([]Node, CompileStatus) {
		if c.Value == nil {
			return positioned("", c.Modifier, "radial")
		}
		if c.Value.Kind != ValueArbitrary || c.Modifier != nil {
			return nil, NotHandled
		}
		return []Node{Decl("--tw-gradient-position", c.Value.Value), image("radial")}, Handled
	}, nil)

	conic := func(c *Candidate, negative bool) ([]Node, CompileStatus) {
		if c.Value == nil {
			if negative {
				return nil, NotHandled
			}
			return positioned("", c.Modifier, "conic")
		}
		if c.Value.Kind == ValueArbitrary {
			if negative || c.Modifier != nil {
				return nil, NotHandled
			}
			return []Node{Decl("--tw-gradient-position", c.Value.Value), image("conic")}, Handled
		}
		named := c.Value.Value
		if !IsPositiveInteger(named) {
			return nil, NotHandled
		}
		angle := named + "deg"
		if negative {
			angle = "calc(" + named + "deg * -1)"
		}
		return positioned("from "+angle, c.Modifier, "conic")
	}
	u.Functional("bg-conic", func(c *Candidate) ([]Node, CompileStatus) { return conic(c, false) }, nil)
	u.Functional("-bg-conic", func(c *Candidate) ([]Node, CompileStatus) { return conic(c, true) }, nil)

	stop := func(name, colorVariable, positionVariable string, stops [][2]string) {
		color := func(value string) ([]Node, CompileStatus) {
			nodes := append(GradientRegistrations(), Decl(colorVariable, value))
			for _, s := range stops {
				nodes = append(nodes, Decl(s[0], s[1]))
			}
			return nodes, Handled
		}
		position := func(value string) ([]Node, CompileStatus) {
			return append(GradientRegistrations(), Decl(positionVariable, value)), Handled
		}
		u.Functional(name, func(c *Candidate) ([]Node, CompileStatus) {
			if c.Value == nil {
				return nil, NotHandled
			}
			if c.Value.Kind == ValueArbitrary {
				value := c.Value.Value
				t := ""
				if c.Value.DataType != nil {
					t = *c.Value.DataType
				} else {
					t = InferDataType(value, []string{"color", "length", "percentage"})
				}
				if t == "length" || t == "percentage" {
					if c.Modifier != nil {
						return nil, NotHandled
					}
					return position(value)
				}
				resolved, ok := AsColor(value, c.Modifier, theme)
				if !ok {
					return nil, NotHandled
				}
				return color(resolved)
			}
			if themeColor, ok := ResolveThemeColor(c, []string{"--background-color", "--color"}, theme); ok {
				return color(themeColor)
			}
			if c.Modifier != nil {
				return nil, NotHandled
			}
			named := c.Value.Value
			if strings.HasSuffix(named, "%") && IsPositiveInteger(named[:len(named)-1]) {
				return position(named)
			}
			return nil, NotHandled
		}, nil)
	}
	stop("from", "--tw-gradient-from", "--tw-gradient-from-position", [][2]string{{"--tw-gradient-stops", GradientStops}})
	stop("via", "--tw-gradient-via", "--tw-gradient-via-position", [][2]string{{"--tw-gradient-via-stops", GradientViaStops}, {"--tw-gradient-stops", "var(--tw-gradient-via-stops)"}})
	stop("to", "--tw-gradient-to", "--tw-gradient-to-position", [][2]string{{"--tw-gradient-stops", GradientStops}})
}
